waspc/waspls/src/Wasp/LSP/SignatureHelp.hs

```diff
@@ -0,0 +1,261 @@
+  ( getSignatureHelpAtPosition,
+    signatureHelpTriggerCharacters,
+    signatureHelpRetriggerCharacters,
+  )
+where
+
+import Control.Lens ((^.))
+import Control.Monad (guard)
+import Control.Monad.Log.Class (MonadLog, logM)
+import Control.Monad.State.Class (MonadState, gets)
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.Maybe (MaybeT (MaybeT, runMaybeT))
+import qualified Data.HashMap.Strict as M
+import Data.List (elemIndex, foldl', intersperse)
+import Data.Maybe (fromMaybe, mapMaybe)
+import Data.String (IsString (fromString))
+import qualified Data.Text as Text
+import qualified Language.LSP.Types as LSP
+import Text.Printf (printf)
+import Wasp.Analyzer.Parser.CST.Traverse (Traversal, fromSyntaxForest)
+import Wasp.Analyzer.Type (Type)
+import qualified Wasp.Analyzer.Type as Type
+import Wasp.LSP.ServerState (ServerState, cst, currentWaspSource)
+import Wasp.LSP.Syntax (lspPositionToOffset, toOffset)
+import Wasp.LSP.TypeInference (ExprKey (Key, List, Tuple), findExprPathAtLocation, findTypeForPath)
+
+-- | Configuration for 'LSP.Options', used in "Wasp.LSP.Server".
+--
+-- When the client types one of these characters, it will send a SignatureHelp
+-- request. It is configured so that signatures will display when starting a
+-- dictionary, list, tuple, or starting a new field/value in the container.
+signatureHelpTriggerCharacters :: Maybe [Char]
+signatureHelpTriggerCharacters = Just "{[(:,"
+
+-- | Configuration for 'LSP.Options', used in "Wasp.LSP.Server".
+--
+-- When a client is already displaying signature help, typing one of these
+-- characters will cause it to update the displayed signature by sending a new
+-- SignatureHelp request. It is configured so that signatures will update when
+-- ending a dictionary, list, or  tuple.
+--
+-- NOTE: 'signatureHelpTriggerCharacters' are also counted as retrigger characters.
+signatureHelpRetriggerCharacters :: Maybe [Char]
+signatureHelpRetriggerCharacters = Just "}])"
+
+-- | Get 'LSP.SignatureHelp' at a position in the wasp file. The signature
+-- displays the type of the "container" that the position is in, if any.
+--
+-- SignatureHelp is usually displayed as a popup near the text you are typing,
+-- and highlights the part of the signature you are writing.
+--
+-- A container is an expression that holds other values, such as dictionaries
+-- and lists.
+--
+-- The parameter field of the signature is used for which part of the container
+-- the position is within, such as a key for a dictionary.
+getSignatureHelpAtPosition ::
+  (MonadState ServerState m, MonadLog m) =>
+  LSP.Position ->
+  m LSP.SignatureHelp
+getSignatureHelpAtPosition position = do
+  src <- gets (^. currentWaspSource)
+  gets (^. cst) >>= \case
+    Nothing ->
+      -- No CST in the server state, can't create a signature.
+      return emptyHelp
+    Just syntax -> do
+      let offset = lspPositionToOffset src position
+      let location = toOffset offset (fromSyntaxForest syntax)
```

Comment:
This reads a bit weird:
1. You get the `offset`
2. Then you use the `toOffset` function to get the location

Is a better name for that function `toLocation`?

Reply:
Sounds better 👍